### PR TITLE
feat(cli): support `Fn::ImportValue` intrinsic function for hotswap deployments

### PR DESCRIPTION
**Pull request by tomwwright**
## Purpose 🎯 

Extend the `EvaluateCloudFormationTemplate` class to support the `Fn::ImportValue` intrinsic function. This allows for more diverse templates to be evaluated for the purposes of determining eligibility for `--hotswap` deployments

Closes https://github.com/aws/aws-cdk/issues/21320

## Approach 🧠 

Implement `LazyLookupExport` in similar fashion to `LazyListStackResources` to cache required CloudFormation API calls _(preference was to implement using a generator function instead so style is not entirely consistent, is this an issue?)_

Add some basic unit tests for `EvaluateCloudFormationTemplate.evaluateCfnExpression()` is they were absent, then add some tests for `Fn::ImportValue`

## Todo 📝 

- [x] Update doco where appropriate
- [x] Add to hotswap deployment tests
- [x] Look for appropriate integration tests to update

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by tomwwright** [comment]
`Exception Request: Features must contain a change to an integration test file and the resulting snapshot.`

I've reviewed the integration tests for `packages/aws-cdk` and I don't see a relevant test to update -- can someone indicate which would be appropriate, or confirm that this `feat` doesn't touch any?

**Comment by tomwwright** [comment]
`Clarification Request: CLI code has changed. A maintainer must run the code through the testing pipeline`

I'm not certain on how to action this one -- could someone help me get this linter request resolved?

File: packages/aws-cdk/lib/api/evaluate-cloudformation-template.ts
Comment (tomwwright) [reply packages/aws-cdk/lib/api/evaluate-cloudformation-template.ts]: Renamed to `cfnExport` as `export` is a keyword _sigh_

File: packages/aws-cdk/lib/api/evaluate-cloudformation-template.ts
Comment (tomwwright) [reply packages/aws-cdk/lib/api/evaluate-cloudformation-template.ts]: Moved to a `const` within the loop, but unable to remove the type def 👇 

```
'response' implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer. ts(7022)
```

File: packages/aws-cdk/README.md
Comment (tmokmss) [inline packages/aws-cdk/README.md:438]: Maybe worth noting that `Fn::GetAtt` is only partially supported? It can only evaluate pre-defined attributes here:

https://github.com/aws/aws-cdk/blob/7df11f4d8475dbf6deefe8fa12df4fea0125b860/packages/aws-cdk/lib/api/evaluate-cloudformation-template.ts#L352-L365

Comment (tomwwright) [reply packages/aws-cdk/README.md:438]: Thanks @tmokmss, will update. Is it reasonable to include a link to those predefined attributes in the code from this user facing doco as context? Or better to just put a general "only partially supported" qualifier against it?

**Comment by tomwwright** [comment]
Have rebased to tip of `main` to resolve conflicts

**Comment by tomwwright** [comment]
Based on changes in https://github.com/aws/aws-cdk/pull/27195 I've refactored away my `lookupExport` prop and handled it as an internal concern to be consistent

File: packages/aws-cdk/README.md
Comment (tmokmss) [reply packages/aws-cdk/README.md]: Thanks! Imo the latter is enough for README.md, since it's intended for end users after all.

Comment (tomwwright) [reply packages/aws-cdk/README.md]: Have corrected

**Comment by tomwwright** [comment]
Hey @tmokmss -- anything I can do to help this one along? We're excited to be able to leverage `--hotswap` within our organisation that is enabled by this PR

**Review by tmokmss — COMMENTED** [review]
A few minor comments. Overall looks good to me, thanks!

**Comment by tmokmss** [comment]
Hi @tomwwright, we need a review from one of the CDK core members. I think we should wait a bit more, as there are still older PRs waiting for reviews.

**Comment by tomwwright** [comment]
> Hi @tomwwright, we need a review from one of the CDK core members. I think we will have to wait a bit more as there are still older PRs waiting for reviews. I'm looking forward to this feature as well :)

Ah thanks @tmokmss, understood :) It's helpful to know that because now I know where I am in the "queue"! Thanks for providing review in the meantime!

File: packages/aws-cdk/lib/api/evaluate-cloudformation-template.ts
Comment (tomwwright) [reply packages/aws-cdk/lib/api/evaluate-cloudformation-template.ts:56]: Good call, we'll just skip over them. At the time I wanted to initially provide the least opinionated implementation

File: packages/aws-cdk/lib/api/evaluate-cloudformation-template.ts
Comment (tomwwright) [reply packages/aws-cdk/lib/api/evaluate-cloudformation-template.ts:62]: I wanted to provide an implementation that performed well under scenarios where a user had many many exports so I wanted to avoid caching all exports up front. But perhaps I can update my implementation to retain the iterator and pick up searching where it left off in subsequent calls

File: packages/aws-cdk/README.md
Comment (vinayak-kukreja) [inline packages/aws-cdk/README.md:436]: Thank you for adding this documentation. Really appreciate this.

File: packages/aws-cdk/README.md
Comment (vinayak-kukreja) [inline packages/aws-cdk/README.md:446]: If possible (Not required for this PR), can you mention what the partial support looks like?

File: packages/aws-cdk/lib/api/evaluate-cloudformation-template.ts
Comment (vinayak-kukreja) [reply packages/aws-cdk/lib/api/evaluate-cloudformation-template.ts:56]: We can skip the [exports](https://docs.aws.amazon.com/AWSCloudFormation/latest/APIReference/API_Export.html) with not `Name`. It would be strange to see such a scenario, have you encountered these?

File: packages/aws-cdk/lib/api/evaluate-cloudformation-template.ts
Comment (vinayak-kukreja) [reply packages/aws-cdk/lib/api/evaluate-cloudformation-template.ts:62]: What would be the benefit of listing exports for each lookup for a hotswap? Do you expect the list to be changed in each lookup?

Comment (vinayak-kukreja) [reply packages/aws-cdk/lib/api/evaluate-cloudformation-template.ts:62]: Would not [this](https://github.com/aws/aws-cdk/pull/27292/files#diff-ff9724092d7f1e6bf3141e51ddb00ecbc2148e16f141b0bd2074d6edc60c16b6R69) list all the exports present?

**Review by vinayak-kukreja — COMMENTED** [review]
Hey, this is great!!

I have just added some comments to get some clarification. Let me know if any of my comments are unclear.

**Comment by vinayak-kukreja** [comment]
Hey, can you take a look into adding CLI integ tests for this feature. This [file](https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk-testing/cli-integ/tests/cli-integ-tests/cli.integtest.ts) is a good starting point.

[Here](https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk-testing/cli-integ/resources/cdk-apps/app/app.js) are some of the defined test stacks.

File: packages/aws-cdk/lib/api/evaluate-cloudformation-template.ts
Comment (tomwwright) [reply packages/aws-cdk/lib/api/evaluate-cloudformation-template.ts:62]: The [linked function](https://github.com/aws/aws-cdk/pull/27292/files#diff-ff9724092d7f1e6bf3141e51ddb00ecbc2148e16f141b0bd2074d6edc60c16b6R69) is a generator function so it will yield pages as requested by the [loop here](https://github.com/aws/aws-cdk/pull/27292/files#diff-ff9724092d7f1e6bf3141e51ddb00ecbc2148e16f141b0bd2074d6edc60c16b6R54) that is searching for the export

The benefit isn't so much about ensuring we have the latest, but more about avoiding cost of iterating all the exports if we don't need to (i.e. if found in the first page then no need to enumerate the rest)

File: packages/aws-cdk/lib/api/evaluate-cloudformation-template.ts
Comment (tomwwright) [reply packages/aws-cdk/lib/api/evaluate-cloudformation-template.ts:56]: I have not encountered that scenario, no. I'll update to skip over exports with no `Name`! 🙌

**Comment by tomwwright** [comment]
> Hey, can you take a look into adding CLI integ tests for this feature. This [file](https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk-testing/cli-integ/tests/cli-integ-tests/cli.integtest.ts) is a good starting point.
> 
> [Here](https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk-testing/cli-integ/resources/cdk-apps/app/app.js) are some of the defined test stacks.

Thanks for the pointer! I'll take a look

File: packages/aws-cdk/README.md
Comment (tomwwright) [reply packages/aws-cdk/README.md:446]: Hey @vinayak-kukreja the best clarification would be a link to [the code](https://github.com/aws/aws-cdk/blob/1929f2e83d14ddcd6123bb9b522456a8c1bd5772/packages/aws-cdk/lib/api/evaluate-cloudformation-template.ts#L477) that implements it but that may not be suitable for this user-facing doco. I could list partially supported resources but that has a chance of becoming stale doco. Thoughts?

cc @tmokmss

**Comment by tomwwright** [comment]
Hey @vinayak-kukreja, have added an integration test as asked (I'm fairly certain there isn't any snapshot associated as per the automated linter message)

Have also implemented skipping over any exports that don't have `Name` (if it ever occurs)

File: packages/aws-cdk/README.md
Comment (vinayak-kukreja) [reply packages/aws-cdk/README.md:447]: Ok, so there can be a link here. For instance, saying `see here[the link embedded] for more information`?

**Comment by vinayak-kukreja** [comment]
Hey, I have added one NIT comment and also executing this PR in our test pipeline for running CLI tests.

File: packages/aws-cdk/README.md
Comment (tomwwright) [reply packages/aws-cdk/README.md:447]: Sure thing, can do!

**Comment by tomwwright** [comment]
@vinayak-kukreja added the doco link and rebased to tip of `main` to update -- let me know if there was anything else pending!

**Comment by vinayak-kukreja** [comment]
Hey apologies for the delay. We are facing some issues with our test pipeline. I will update here as I get a chance to run this PR through the pipeline.

**Review by vinayak-kukreja — APPROVED** [review]
Hey @tomwwright , the pipeline run was successful. Thank you for contributing to the cdk. Really appreciate your effort.

**Comment by tomwwright** [comment]
Thank you for your review @vinayak-kukreja 🙌

**Pull request merged**
